### DISCUSSION TITLE
Composable Models

### DISCUSSION THREAD

**Discussion by gnodar01** [broadinstitute/superurop-log#8] [Ideas]
Over the last week or so [we've been discussing](https://github.com/broadinstitute/superurop-log/discussions/7#discussioncomment-7579909) going forward with composable models. The rough idea is to have individual components of an instance segmentation model, that are pre-trained on some task such as imagenet or coco. One component might be an object detector (probably single-shot such as yolo or ssd) which outputs boundbox coordinates (and classes, but we may not even need those). Another component might be a semantic segmenter, which takes in crops defined by the bounding boxes of the object detector, and semantically segments into bg/fg.

The hypothesis is, given pretrained components of the sort described above, can we add on a relatively slim number of layers on each, to have a fine-tunable model?

If so, we can have a composable pipline that goes `images -> object detector (pretrained, inference only) -> slim layers (trainable) -> output bbox coords -> image crops -> semantic segmenter (pretrained, inference only) -> slim layers (trainable) -> instances`

In the end this pipeline would be deployed to the browser.

The object detector and semantic segmenter would be pretrained in python on a laptop or DGX, and then converted to a tfjs graph model. We have determined that is possible with more or less arbitrary tensorflow using [tfjs convertor](https://github.com/tensorflow/tfjs/tree/master/tfjs-converter) or pytorch models using the [ultralytics Exporter class](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/exporter.py#L127) (which does `pytorch -> onnx -> tensorflow saved model -> tfjs converter -> tfj graph model`).

The slim layers would be tfjs layers models, and therefore trainable in-browser.

For the purposes of validating the hypothesis first however, we'll do it all in python, as a proof of concept. We can worry about converting to tfjs and verifying its performance in the browser after.

**Comment by gnodar01** [comment]
We discussed Mask-RCNN style (two pass) vs SSD (one pass) vs yolo (one pass) for object detection, and are leaning towards yolo. Interestingly [yolo has gone through a number of iterations](https://sandar-ali.medium.com/brief-timeline-of-yolo-models-from-v1-to-v8-7d383140afe2) improved substantially over the past few years ([review](https://arxiv.org/pdf/2304.00501.pdf)). v1-v3 had the original author Joseph Redmon working on it, [but has since left CV research](https://twitter.com/pjreddie/status/1230524770350817280?lang=en). Alexey Bochkovskiy worked on v4. Other teams, including ultralytics, have been improving it up to the current day v8. Ultralytic's [v5](https://docs.ultralytics.com/yolov5/) is particularly interesting because it introduces multi-task learning, including segmentation. This continues with ultralytics [v8](https://docs.ultralytics.com/quickstart/), which take into account improvements in efficiency made in [v6](https://github.com/meituan/YOLOv6) such as quantization, etc. Its also a generally nice framework, with the previously mention export functionality.

The v8 family of models for the task of instance segmentation are called [YOLOv8-seg](https://docs.ultralytics.com/tasks/segment/). [By this comment](https://github.com/ultralytics/ultralytics/issues/1710) it seems that yolov8-seg architecture is pretty much the same as the v8 object detector with "an extra output module in the head which outputs masks coefficients and an added FCN layer", which is inspired by the [YOLACT](https://arxiv.org/pdf/1904.02689.pdf) paper. Worth reading is the [Fast Segment Anything](https://arxiv.org/pdf/2306.12156.pdf) paper which uses yolov8-seg and references YOLACT (and is also [available](https://docs.ultralytics.com/models/fast-sam/) via the ultralytics library).

**Reply by gnodar01** [reply]
With yolov8-seg, the question is how do we disentangle it to fit the component pipeline described above? Or alternatively, should we? Maybe we can use the entire model, pretrained on some task, rip out its current classes, and add some relatively "slim" trainable layers on top to fine tune the bbox coordinates + masks for the task we care about?

**Comment by gnodar01** [comment]
@eddogola feel free to correct me on anything, and add any additional points here. Once you're ready to write some code, press "Create issue from discussion" and convert it into an issue, since this is the next immediate chunk of work we'll be working on.

**Comment by bethac07** [comment]
Questions around picking the object detector head:
- One stage vs two
- Proposal vs proposal-free
- Anchor vs anchor-free

(We may not get to investigating this but may want to eventually get into that)

**Comment by bethac07** [comment]
Possible object detectors:
- Yolov1 (one stage, anchor free, but bad recall because no FPN)
- some other Yolo (eg most recent v8, which is anchor and proposal free)
- RPN (as sub-component for proposal-based methods, ie two stage models)
- FPN (as sub-component)
- SSD (anchor-based, single stage)
- FCOS (one stage, anchor free, fully convolutional, good recall via FPN)
Others?
